drivers/net/mlx5/mlx5_vdpa.c

```diff
@@ -163,6 +200,122 @@ static int mlx5_vdpa_release_rx(struct vdpa_priv *priv)
 	return 0;
 }
 
+static
+struct mlx5_devx_mkey *mlx5_vdpa_create_mkey(struct ibv_context *ctx,
+					struct mlx5_devx_mkey_attr *mkey_attr)
+{
+	uint32_t in[MLX5_ST_SZ_DW(create_mkey_in)] = {0};
+	uint32_t out[MLX5_ST_SZ_DW(create_mkey_out)] = {0};
+	uint32_t status;
+	void *mkc;
+	struct mlx5_devx_mkey *mkey = NULL;
+
+	mkey = rte_zmalloc("mkey", sizeof(*mkey), 64);
+
+	MLX5_SET(create_mkey_in, in, opcode, MLX5_CMD_OP_CREATE_MKEY);
+	mkc = MLX5_ADDR_OF(create_mkey_in, in, memory_key_mkey_entry);
+	MLX5_SET(mkc, mkc, lw, 0x1);
+	MLX5_SET(mkc, mkc, lr, 0x1);
+	MLX5_SET(mkc, mkc, rw, 0x1);
+	MLX5_SET(mkc, mkc, rr, 0x1);
+	MLX5_SET(mkc, mkc, access_mode_1_0, MLX5_MKC_ACCESS_MODE_MTT);
+	MLX5_SET(mkc, mkc, qpn, 0xffffff);
+	MLX5_SET(mkc, mkc, length64, 0x0);
+	MLX5_SET(mkc, mkc, pd, mkey_attr->pd);
+	MLX5_SET(mkc, mkc, mkey_7_0, 0x50);//FIXME: should be dynamic
+	MLX5_SET(mkc, mkc, translations_octword_size,
+		(((mkey_attr->size + 4095) / 4096) + 1) / 2);
+	MLX5_SET(create_mkey_in, in, translations_octword_actual_size,
+		(((mkey_attr->size + 4095) / 4096) + 1) / 2);
+	MLX5_SET(create_mkey_in, in, pg_access, 1);
+	MLX5_SET64(mkc, mkc, start_addr, mkey_attr->addr);
+	MLX5_SET64(mkc, mkc, len, mkey_attr->size);
+	MLX5_SET(mkc, mkc, log_page_size, 12);
+	MLX5_SET(create_mkey_in, in, mkey_umem_id, mkey_attr->pas_id);
+
+	mkey->obj = mlx5_glue->dv_devx_obj_create(ctx, in, sizeof(in), out,
+					   sizeof(out));
+	if (!mkey->obj) {
+		DRV_LOG(ERR, "Can't create mkey error %s\n", strerror(errno));
+		goto error;
+	}
+	status = MLX5_GET(create_mkey_out, out, status);
+	mkey->key = MLX5_GET(create_mkey_out, out, mkey_index);
+	mkey->key = (mkey->key << 8) | 0x50;
+	DRV_LOG(DEBUG, "create mkey status %d mkey value %d\n",
```

Comment:
if possible - remove the \n (later when running on DEBUG it creates an redundant empty line)